### ISSUE TITLE
Support for NX monorepos continues from older discussions

### ISSUE THREAD

**Issue by ryanbas21** [changesets/changesets#1188]
There have been a list of issues over time that have referenced the desire for changesets to support NX style monorepos. 

Issues in the past: 
#801 
#849 
#773 
#761 

Probably a few others.

### How does NX differ
Nx by default will have a `dist` directory in the workspace root. So a typical monorepo may have a file structure in this shape (not prescriptive, and just a rough outline):

```
 - nx.json
 - package.json [root, often times all dependencies are here]
 - dist 
 - packages/
     - my-package-a
        - package.json
        - CHANGELOG.md
        -  README.md
        - src
      -my-package-b
        - package.json
        - CHANGELOG.md
        -  README.md
        - src
```

While workspaces are not required in `NX` monorepos, they are supported if you choose to use them. This is configured in the root of the workspace dependent on your package manager for how.

When NX builds - often times the `package.json` from the package itself, is copied, or generated and then published into the `dist` folder.  NX also will typically analyze libraries dependencies and automatically compiles the dependencies in the resulting package.json file (if configured with (updateBuildableProjectDepsInPackageJson)[https://nx.dev/packages/angular/executors/package#updatebuildableprojectdepsinpackagejson] ) 

So `dist` will look like so

```
dist
   packages/
      my-package-a
        package.json
        index.mjs
        index.cjs
        index.d.cts
        index.d.mts
        README.md [ this file isn't always copied, the repo would configure its own movement of this file typically through assets ] 
```

Publishing a package is of course running npm publish with this directory or just moving into this directory and publishing.

This presents a few challenges that I see (I'm unfamiliar with the changesets repo itself and how it works so this is just what I see from an agnostic standpoint).

### Challenges
 - the package.json version would need to be updated or be able to be updated in the src. modifying the dist/ package.json is important for the release, but if we are able to _only_ modify the src before building, then nx should be able to copy the current version that was just modified. If not, we would need to update the package.json in both `dist/packages/my-package/package.json` and also `packages/my-package/package.json`
 
 - The same would go for the CHANGELOG.md file. If we are able to run these modifications _before_ building and publishing, NX will be able to be configured to copy these over if configured by the user.
 
 - NX supports the concept of `affected` tasks. `affected` tasks are simply NX determining what files/packages / libraries /apps have changed (or been affected by code) and rebuilding the necessary packages. This is determined by NX's project graph. `affected` commands are any command that can be cached (as defined in your project) so it is possible to do `affected` releases. I am not sure if this presents any issues unless your in a `fixed` or `linked` configuration and you are then modifying a CHANGELOG or package.json, in which case, those packages just need to get rebuilt. As long as the actual `packages/[my-package]/ is changed (affected), these would be rebuilt by NX when the `build` task is dispatched. 
 
 - NX does not enforce a workspace configuaration. It's probably not unreasonable to ask a `workspace` configuration be made if you want to use changesets, but the only other way to determine the possible packages to publish is to read the monorepos setup.
 
 You could do this via the `tsconfig.base.json` where each `path` is essentially an alias to a possible app/lib/package/etc. Any package which has a package.json could be deemed "publishable".  A lot of this feels flakey as it is dependent on us digging into NX's implementation details which feasibly could change.
 
 
 ### Wrapping up thoughts
 
 Since my understanding of the changesets repo isnt very high i can't offer at this time much into _how_ to attack the problem. However as i've touched on, the simpest solution seems to be to run `changeset version` before building. If you do that, and are able to point changesets at the proper directories, then reasonably, we could update the package.json / changelog's, whatever else, and then ask `NX` to build (via a provided build command).
 
 After we build, changesets would need to be told where to look (typically this is `${workspaceRoot}/dist/packages/*` but could be named whatever and can't be guaranteed what naming is.
 
 @Andarist let me know if this outline helps fill in some gaps with Nx and Changesets and I can provide more info. Also happy to hear from anyone else who has more insight into these two tools and the gaps that exist.

**Comment by Andarist** [comment]
> the package.json version would need to be updated or be able to be updated in the src. modifying the dist/ package.json is important for the release, but if we are able to only modify the src before building, then nx should be able to copy the current version that was just modified. If not, we would need to update the package.json in both dist/packages/my-package/package.json and also packages/my-package/package.json

This is correct. Changesets have separate `changeset version` and `changeset publish`. So by the time we get to publishing all `package.json` files in workspaces should already be updated so Nx should be able to just copy the updated version. 

> NX supports the concept of affected tasks. affected tasks are simply NX determining what files/packages / libraries /apps have changed (or been affected by code) and rebuilding the necessary packages. This is determined by NX's project graph. affected commands are any command that can be cached (as defined in your project) so it is possible to do affected releases. I am not sure if this presents any issues unless your in a fixed or linked configuration and you are then modifying a CHANGELOG or package.json, in which case, those packages just need to get rebuilt. As long as the actual packages/[my-package]/ is changed (affected), these would be rebuilt by NX when the build` task is dispatched.

Changesets are using a somewhat easier approach here. We just check if workspaces are published - so we check all `package.json` files and query the npm registry to check of that given version is already there. If it's not... we try to publish it.

While maybe this isn't optimal, it works reasonably well. You don't often try to publish things anyway. Nx could maybe inform us about affected packages etc as an optimization - but likely that's not even needed and I wouldn't bother with this right away. 

> NX does not enforce a workspace configuaration. It's probably not unreasonable to ask a workspace configuration be made if you want to use changesets, but the only other way to determine the possible packages to publish is to read the monorepos setup.

This I don't fully understand. How does it even behave if there is no workspace configuration? Is every directory with `package.json` in it considered to be a workspace in such a case?

> After we build, changesets would need to be told where to look (typically this is ${workspaceRoot}/dist/packages/* but could be named whatever and can't be guaranteed what naming is.

Is there any specific Nx command that you execute to publish packages? Perhaps it already does the heavy-lifting of finding the publishable directory for a given workspace?

**Comment by ryanbas21** [comment]
> This I don't fully understand. How does it even behave if there is no workspace configuration? Is every directory with package.json in it considered to be a workspace in such a case?

Each project with Nx has a project.json, which is akin to an angular.json, but its how NX runs tasks from each project, its agnostic to running in workspaces or not. NX prefers but doesn't require, a singular package.json for dependencies but any publishable library would have its own package.json (although may list no dependencies). NX when it builds, calculates and updates the package.json so the artifact package.json would have that information as needed.

I don't know the inner details of NX super well, but i'm guessing that project.json file is how NX is determining if its a "workspace-like". There used to be a workspace.json which listed all of the workspaces in the project but that has been removed in newer versions of NX.

I believe the key to how NX determines everything is leveraged from the project graph, and its probably the core feature for everything NX does. 

> Is there any specific Nx command that you execute to publish packages? Perhaps it already does the heavy-lifting of finding the publishable directory for a given workspace?

Nope. NX doesn't give you anything that prescribes how you publish. Someone could add a "publish" task to the project.json for a library, but chances are thats just calling some form of `npm publish` under the hood. Theres no "NX" specific way of publishing and early videos of theres explain how to publish an npm package by `cd dist/packages/my-package` an calling `npm publish`.
